scripts/irods/test/test_misc.py

```diff
@@ -198,47 +198,85 @@ def test_server_removes_leftover_temporary_rulebase_pid_files_on_startup__issue_
-                                    'delay("<INST_NAME>irods_rule_engine_plugin-irods_rule_language-instance</INST_NAME>") {{ msiModAVUMetadata("-d", "{0}", "add", "{1}", "v1", "u2") }}'.format(object_name, attribute),
-                                    'null',
-                                    'ruleExecOut'])
-
-        # Wait for the delay server to process the rule
-        time.sleep(45)
-        self.admin.assert_icommand('imeta ls -d %s' % (object_name), 'STDOUT_SINGLELINE', ['attribute: ' + attribute])
-        self.admin.assert_icommand(['irm', "-f", 'hello'])
+
+        try:
+            # Find the main server process
+            for proc in psutil.process_iter():
+                # It won't have any parents.
+                if proc.name() == 'irodsServer' and proc.parent().name() != 'irodsServer':
+                    main_server = proc
+                    break
+
+            # It's possible that the server was never found, which would be a problem.
+            self.assertIsNotNone(main_server)
+
+            # Kill the agent factory and the delay server
+            for child in main_server.children(recursive=True):
+                child.kill()
+
+            # If the API is up after this it seems it should be entirely working.
+            # run_icommand returns (stdout, stderr, returncode), so try until itouch returns 0.
+            lib.delayAssert(lambda: self.admin.run_icommand(['itouch', object_name])[2] == 0)
+
+            def process_has_at_least_two_child_processes_with_expected_names(proc):
+                children = proc.children()
+                # debugging print
+                print(children)
+                for c in children:
+                    # debugging print
+                    print(f'child name [{c.name()}]')
+                    if c.name() != 'irodsDelayServer' and c.name() != 'irodsServer':
+                        return False
+                return len(children) >= 2
+
+            # Make super-extra sure that both the delay server and the agent factory are up.
+            # This check is necessary because it has been observed that the name of child
+            # processes can be truncated and then later respawned with the correct name.
+            lib.delayAssert(lambda:
+                    process_has_at_least_two_child_processes_with_expected_names(main_server))
+
+            # Double-check that there is exactly one delay server and exactly one agent factory
+            # instance. As the agent factory process shares a name with the main server process
+            # as well as iRODS agent processes, we check for the 'irodsServer' process name.
+            seen_process_names = defaultdict(int)
+            for i in main_server.children():
+                seen_process_names[i.name()] += 1
+            self.assertEqual(1, seen_process_names['irodsDelayServer'])
+            # TODO: Is it possible that a connection can open to spawn an agent (timing issue)?
```

Comment:
yes, it is - you could make the 'wake up interval' really big at the beginning of this test... reducing the window/chance?

Reply:
We may need to wrap this in a `delayAssert`. If we widen the sleep window, it *could* still occur (plus the test would take way longer to complete), but if we just check again via `delayAssert`, there's an almost 0 chance that there will be extra child processes *100 times* unless something is actually wrong. Thoughts?

Reply:
your assessment seems correct to me.

Reply:
Cool. Done